# Exploratory Data Analysis: Global Literacy Rates

## Project Overview

This notebook explores the Global Literacy dataset from Our World in Data.

The dataset contains historical and contemporary estimates of literacy rates across countries and regions, spanning several centuries. Before using the data in a larger analysis, it is important to understand its structure, coverage, quality, and limitations.

This notebook focuses exclusively on exploratory data analysis (EDA) of the literacy dataset.

---

## Objectives

The goals of this analysis are to:

- Understand how literacy is defined and measured.
- Examine the structure of the dataset.
- Assess data quality and completeness.
- Investigate geographic and temporal coverage.
- Identify potential limitations and caveats.
- Generate meaningful analytical questions for future research.

---

## About the Dataset

### Definition of Literacy

According to UNESCO, literacy is defined as:

> The ability of a person aged 15 years and older to read, understand, and write a short, simple statement about their everyday life.

The literacy rate represents the percentage of the adult population (15+) that meets this definition.

---

## Why This Dataset Matters

Literacy is a foundational indicator of educational development and human capital. Understanding literacy trends can provide insight into broader social and economic outcomes, including education access, workforce development, and poverty reduction.

---

## EDA Roadmap

This notebook follows a structured exploratory data analysis process:

### 1. Documentation Review

- Review dataset metadata
- Understand variables and definitions
- Identify known limitations

### 2. Data Inspection

- Examine dataset structure
- Review column data types
- Check missing values and duplicates

### 3. Coverage Analysis

- Explore country coverage
- Explore year coverage
- Identify data availability patterns

### 4. Question Generation

- Identify meaningful analytical questions
- Assess suitability for future research projects

### 5. Summary of Findings

- Document key observations
- Identify next steps for analysis

### file imports And Data

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [7]:
df = pd.read_csv('literate-and-illiterate-world-population.csv')

# Documentation Review

In [20]:
df.head()

,Entity,Code,Year,Illiterate,Literate
0,Afghanistan,AFG,1950,97.000000,3.00000
1,Afghanistan,AFG,1979,82.000000,18.00000
2,Afghanistan,AFG,2011,69.000000,31.00000
3,Afghanistan,AFG,2015,66.246155,33.75384
4,Afghanistan,AFG,2021,63.000000,37.00000


the 2 columns over here 'Illiterate' and 'Literate' are based on percentage

In [19]:
df.columns.to_frame(index=False)

,0
0,Entity
1,Code
2,Year
3,Illiterate
4,Literate


In [21]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2059 entries, 0 to 2058
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Entity      2059 non-null   str    
 1   Code        1687 non-null   str    
 2   Year        2059 non-null   int64  
 3   Illiterate  2059 non-null   float64
 4   Literate    2059 non-null   float64
dtypes: float64(2), int64(1), str(2)
memory usage: 80.6 KB


In [22]:
df.describe()

,Year,Illiterate,Literate
count,2059.000000,2059.000000,2059.000000
mean,1986.084507,27.421922,72.578078
std,64.008447,24.485748,24.485748
min,1475.000000,0.000000,0.000000
25%,1983.000000,6.704529,57.981690
50%,2002.000000,21.000000,79.000000
75%,2013.000000,42.018310,93.295470
max,2023.000000,100.000000,100.000000


In [25]:
df.shape

(2059, 5)

2059 rows and 5 cols

### Find missing values and Null values

In [27]:
df.isna().sum()

Entity          0
Code          372
Year            0
Illiterate      0
Literate        0
dtype: int64

In [36]:
df.isnull().sum()

Entity          0
Code          372
Year            0
Illiterate      0
Literate        0
dtype: int64

interesting we have 372 null values and missing values in the column "code" which is the shrotened version of teh country

### duplicates 

In [28]:
df.duplicated().sum()

np.int64(0)

In [34]:
df.duplicated(subset=['Entity', 'Year']).sum()

np.int64(0)

In [29]:
df[['Literate', 'Illiterate']].describe()

,Literate,Illiterate
count,2059.000000,2059.000000
mean,72.578078,27.421922
std,24.485748,24.485748
min,0.000000,0.000000
25%,57.981690,6.704529
50%,79.000000,21.000000
75%,93.295470,42.018310
max,100.000000,100.000000


In [30]:
(df['Literate'] + df['Illiterate']).head()

0    100.000000
1    100.000000
2    100.000000
3     99.999995
4    100.000000
dtype: float64

there are some rows that does not complete 100%

In [32]:
(df['Literate'] + df['Illiterate']).head()

0    100.000000
1    100.000000
2    100.000000
3     99.999995
4    100.000000
dtype: float64

In [47]:
(df['Literate'] + df['Illiterate']).describe()

count    2059.000000
mean      100.000000
std         0.000001
min        99.999995
25%       100.000000
50%       100.000000
75%       100.000000
max       100.000004
dtype: float64

### Check the coverage

In [45]:
# min and max year rows
df.loc[[df['Year'].idxmin(),
         df['Year'].idxmax()]
         ]

,Entity,Code,Year,Illiterate,Literate
84,Belgium,BEL,1475,90.0,10.0
55,Azerbaijan,AZE,2023,0.0,100.0


Earliest year

In [ ]:
earliest_year = df['Year'].min()
latest_year = df['Year'].max()

earliest_data = df[df['Year'] == earliest_year]
latest_data = df[df['Year'] == latest_year]

In [59]:
# earlest data
earliest_data

,Entity,Code,Year,Illiterate,Literate
84,Belgium,BEL,1475,90.0,10.0
566,France,FRA,1475,94.0,6.0
595,Germany,DEU,1475,91.0,9.0
717,Ireland,IRL,1475,100.0,0.0
725,Italy,ITA,1475,85.0,15.0
1232,Netherlands,NLD,1475,83.0,17.0
1448,Poland,POL,1475,100.0,0.0
1638,Spain,ESP,1475,97.0,3.0
1770,Sweden,SWE,1475,99.0,1.0
1871,United Kingdom,GBR,1475,95.0,5.0


In [53]:
(df['Year'] == earliest_year).sum()

np.int64(11)

### Initial Observation
there are only 11 rows in where the year is 1475

In [66]:
earliest_data['Entity']

84                                      Belgium
566                                      France
595                                     Germany
717                                     Ireland
725                                       Italy
1232                                Netherlands
1448                                     Poland
1638                                      Spain
1770                                     Sweden
1871                             United Kingdom
1980    Western Europe (Buringh and van Zanden)
Name: Entity, dtype: str